pytlu/tlu_eudaq.py

```diff
@@ -0,0 +1,405 @@
+                if actual_trigger_number != self.last_trigger_number + 1:
+                    logging.warning('Expected != Measured trigger number: %d != %d', self.last_trigger_number + 1, actual_trigger_number)
+                self.last_trigger_number = actual_trigger_number
+                self.callback(data=data, skipped_triggers=skipped_triggers, event_counter=self.event_counter)
+                self.event_counter += 1
+
+
+def replay_tlu_data(data_file, real_time=True):
+    '''
+    Replay data from file.
+
+    Parameters
+    ----------
+    real_time: boolean
+        Delays return if replay is too fast to keep
+        replay speed at original data taking speed.
+    '''
+
+    with tb.open_file(data_file, mode="r") as in_file_h5:
+        meta_data = in_file_h5.root.meta_data[:]
+        raw_data = in_file_h5.root.raw_data
+        n_readouts = meta_data.shape[0]
+
+        last_readout_time = time.time()
+
+        last_trigger_number = -1
+
+        for i in tqdm(range(n_readouts)):
+            # Raw data indeces of readout
+            i_start = meta_data['index_start'][i]
+            i_stop = meta_data['index_stop'][i]
+
+            t_start = meta_data[i]['timestamp_start']
+
+            # Determine replay delays
+            if i == 0:  # Initialize on first readout
+                last_timestamp_start = t_start
+            now = time.time()
+            delay = now - last_readout_time
+            additional_delay = t_start - last_timestamp_start - delay
+            if real_time and additional_delay > 0:
+                # Wait if send too fast, especially needed when readout was
+                # stopped during data taking (e.g. for mask shifting)
+                time.sleep(additional_delay)
+            last_readout_time = time.time()
+            last_timestamp_start = t_start
+
+            actual_data = raw_data[i_start:i_stop]
+
+            skipped_triggers = meta_data[i]['skipped_triggers']
+            for dat in actual_data:
+                    actual_trigger_number = dat['trigger_id']
+                    trg_number, trg_timestamp = dat['trigger_id'], dat['time_stamp']
+                    # Check for jumps in trigger number
+                    if actual_trigger_number != last_trigger_number + 1:
+                        logging.warning('Expected != Measured trigger number: %d != %d', last_trigger_number + 1, actual_trigger_number)
+                    last_trigger_number = actual_trigger_number
+
+                    yield trg_number, trg_timestamp, skipped_triggers
+
+
+def main():
+    tlu = None
+    input_ch = ['CH0', 'CH1', 'CH2', 'CH3']
+    output_ch = ['CH0', 'CH1', 'CH2', 'CH3', 'CH4', 'CH5', 'LEMO0', 'LEMO1', 'LEMO2', 'LEMO3']
+
+    def th_type(x):
+        if int(x) > 31 or int(x) < 0:
+            raise argparse.ArgumentTypeError("Threshold is 0 to 31")
+        return int(x)
+
+    parser = argparse.ArgumentParser(usage="pytlu -ie CH0 -oe CH0",
```

Comment:
All this is really needed 2 times? Looks like repeating code.

Reply:
You are right, all the argument parsing (except the parsing of the EUDAQ related arguments) is repeated code. We could make a function in tlu.py and then use the function here. What do you think?

Reply:
Yes